// automacao-core-playwright/rollup.config.js
import resolve from '@rollup/plugin-node-resolve';
import babel from '@rollup/plugin-babel';
import json from '@rollup/plugin-json';
import replace from '@rollup/plugin-replace';
import nodePolyfills from 'rollup-plugin-node-polyfills';

export default {
  input: 'src/index.d.ts',
  output: {
    file: 'dist/bundled.js',
    format: 'umd',
    name: 'automacao-core-playwright',
  },
  plugins: [
    resolve({
      preferBuiltins: false, // Desabilita o uso de módulos internos do Node.js
    }),
    json(), // Adiciona suporte para arquivos JSON
    babel({ // Utiliza o Babel para transpilar o código
      babelHelpers: 'bundled',
      presets: [
        ['@babel/preset-env', {
          'targets': '> 0.25%, not dead' // ou você pode especificar os ambientes que você quer suportar
        }]
      ],
    }),
    replace({
      'eval(': 'void(', // Substitui o uso de eval por algo inofensivo ou uma alternativa
      delimiters: ['', ''],
      preventAssignment: true,
    }),
    nodePolyfills(), // Adicione o plugin à lista de plugins
  ],
  acorn: {
    allowReserved: true // Permite que o Rollup use uma versão mais recente do Acorn
  },
  external: [], // Lista de módulos a serem mantidos externos
};
